core/src/main/scala/scala/macros/core/Universe.scala

```diff
@@ -92,5 +98,18 @@ trait Universe {
   // Expansion
   // =========
   type Expansion
+  def enclosingPosition: Position
+  def enclosingOwner: Symbol
+}
+
+trait Input extends Any {
+  def path: Path
+}
```

Comment:
I don't quite understand the concept `Input`, is this concept necessary?

Reply:
Input == SourceFile, it's essentially something that can be read and parsed into trees. Inputs are not always source files, they can be virtual/in-memory, etc. Other methods that could be exposed are for example `contents: Array[Char]`. I added it here since this API is used here https://github.com/lihaoyi/sourcecode/blob/420bea7941d3219e2f1200b14b11010843aea39c/sourcecode/shared/src/main/scala/sourcecode/SourceContext.scala#L130-L141

Reply:
If `Input` is not always files, then the method `def path: Path` looks a little weird.

Reply:
Is `label` a better name?

Reply:
It depends what we are going to expose as the user-facing API. If `def path: Path` is the only API we are going to support, then the abstraction seems can be avoided.

I'm a little uncertain about whether it's good to expose `def content: Array[Char]` to support getting source code by a range. With such APIs, the macro system makes more promises to macro authors, but (1) we know positions of source code are very tricky; (2) it seems most macros don't require this capability.

Reply:
>  (2) it seems most macros don't require this capability.

Do you have data to back that claim? My observations are quite the opposite, macros are heavily used to expose position/source information. I aggregated the most used macros from a corpus of 1M LOC in the community build,https://gist.github.com/olafurpg/baa50ea0fcc2b91c65cdd7da623964fc The top macro accounts for almost 50% of all macro usages 

```
4845: _root_.org.scalactic.source.Position.here()Lorg/scalactic/source/Position;.
```

If we expose `path` but not `contents` then macro libraries will simply read the source file contents from I/O, which is not desirable. We can decide not to expose `path`, but in that case we make it impossible to implement the most used macro in the community build.

From my point of view, `Input`/`SourceFile` is an important concept and should not be conflated with `Position`.

Reply:
I think getting current line & file name (path?) is a justified usage, and it's relatively easy to support them by backends without problem, and user-facing semantics are clear (forget recursive macros for now).

My question is more about should we give macro authors more power: that's to query range of trees thus potentially to retrieve source code by range. The data in the gist doesn't show that this is a common usage.

Note that, in the macro system, there is going to be two different usage Positions:

- `enclosingLocation` that eventually exposes line numbers, files to end Scala users.
- `tree.pos`: so far, the only usage of this is for error reporting service in backend, which usually takes a position. It's not exposed to end Scala users.

For the first one, it's a promise to be fulfilled by backends, and it's fine to have a case class for that in the macro system. For the second usage, we don't need to expose any API on `Pos`, as there's no need to know what they are.

But I'm fine to keep it as an issue, and merge this PR for now.

Reply:
I opened #31, we can continue our discussion there.